### ISSUE TITLE
Torchgeo datamodules override augmentations

### ISSUE THREAD

**Issue by calebrob6** [microsoft/torchgeo#2984]
### Description

If I write a DataModule like below, then `train_aug` will be set to `None` after initialization.

### Steps to reproduce

```
class ExampleDataModule(NonGeoDataModule):

    def __init__(
        self,
        batch_size: int = 64,
        num_workers: int = 0,
        **kwargs: Any,
    ) -> None:
        self.train_aug = K.AugmentationSequential(
            ...
            data_keys=None,
        )
        super().__init__(MyCoolDataset, batch_size, num_workers, **kwargs)
```

(vs. calling `super()` ... immediately)

As a user I would expect `train_aug` to not be overridden.

Further it isn't clear what exactly should and shouldn't be a kwarg.

### Version

latest

**Comment by adamjstewart** [comment]
> As a user I would expect `train_aug` to not be overridden.

I agree that this is confusing. Not a high priority bug because there's an easy workaround (call `super()` first), but still not the best design pattern. Perhaps we could instead use properties? Something like:
```python
@property
def train_aug(self) -> nn.Module:
    return K.AugmentationSequential(...)
```
I didn't do this originally because the average non-CS person (e.g., past me) gets scared when they need to make a subclass, let alone figure out properties, but with the right documentation, this shouldn't be too hard to figure out.

P.S. I've also been thinking about the fact that `val_aug`/`test_aug`/`predict_aug` are almost always identical. I've been considering switching to only two parameters: `preprocess` (all splits) and `augmentation` (only train). This could be timed with a switch from Kornia to torchvision (if we decide to do this) to avoid a backwards-incompatible change.

> Further it isn't clear what exactly should and shouldn't be a kwarg.

Docstrings are your friend. These are passed to the dataset class and include all dataset parameters. Could probably be better documented [here](https://torchgeo.readthedocs.io/en/stable/tutorials/contribute_datamodule.html) if anyone wants to open a PR.

**Comment by isaaccorley** [comment]
I like this but I also think we should all augs to be passed as arguments to the datamodule. This allows for hyperparam sweeps when trying out different augs.

Otherwise when you want to change the augs you have to modify your script every time

**Comment by adamjstewart** [comment]
Yeah, that's the other thing we've debated for a long time. Is there an easy way to store Kornia or torchvision augs in a YAML file? Also want to ensure reproducibility.

**Comment by isaaccorley** [comment]
> Yeah, that's the other thing we've debated for a long time. Is there an easy way to store Kornia or torchvision augs in a YAML file? Also want to ensure reproducibility. 

Yup you can do this easily with hydra since you can write the list of transforms directly in YAML and use hydra.utils.instantiate to create them at runtime. This is what my current experiment workflow is. Also makes it easier to track what exact augs were used for each experiment if you save the config file with your run.

**Comment by adamjstewart** [comment]
I want to keep things simple and use the LightningCLI. Has Lightning added support for hydra yet?

**Comment by adamjstewart** [comment]
Also note that omegaconf/hydra haven't had a new release in almost 3 years. Unclear whether or not they are still maintained.

**Comment by isaaccorley** [comment]
Seems like they are still merging PRs for bug fixes and from dependendabot though. It's pretty stable imo so probably making major changes isn't a priority. I'll ask if they plan to make a new patch release.

**Comment by adamjstewart** [comment]
There are several open issues asking if they plan to make a new release. They have been open for years...

**Comment by calebrob6** [comment]
I'd prefer to put the augs in the Trainers and call `self.aug(batch)` directly in train_step, validation_step, etc. Reasoning:
- Having the actual augmentations applied in an "after batch transfer" function in the datamodule is unintuitive.
- I just ran into a situation where I want to apply an additional augmentation to the model predictions before computing loss / metrics. This will have to be done in the various *_step functions in the trainer anyway.
- I think more generally, just having to override a trainer for custom experiments is better than having to override a trainer and a datamodule

**Comment by adamjstewart** [comment]
> Having the actual augmentations applied in an "after batch transfer" function in the datamodule is unintuitive.

[LightningDataModule](https://lightning.ai/docs/pytorch/stable/data/datamodule.html) is intended to "Apply transforms (rotate, tokenize, etc…)". [LightningModule](https://lightning.ai/docs/pytorch/LTS/common/lightning_module.html) is not. I couldn't find a single PyTorch Lightning tutorial that suggests putting the transforms in the LightningModule instead of the LightningDataModule.

Preprocessing transforms (e.g., normalization, ordinal mapping) tend to be more dependent on the data, not on the model. We should be converting all of our data to be compatible with our models, not the other way around. Data augmentations (e.g., crop, rotate, flip) tend to be independent of both the data and model (I think). Note the "tend", I'm sure there are counterexamples here.

> I just ran into a situation where I want to apply an additional augmentation to the model predictions before computing loss / metrics. This will have to be done in the various *_step functions in the trainer anyway.

This is a better motivation, we often want to undo augmentations for metric computation or visualization. I agree that this is possible but annoying with LightningDataModule. We currently create a new augmentation to undo normalization for visualization. We should probably also do this for metric computation.

> I think more generally, just having to override a trainer for custom experiments is better than having to override a trainer and a datamodule

This is a matter of opinion. Someone who needs to override the LightningDataModule anyway but not the LightningModule would have the exact opposite opinion.

**Comment by calebrob6** [comment]
> I couldn't find a single PyTorch Lightning tutorial that suggests putting the transforms in the LightningModule instead of the LightningDataModule.

augmentation in a LightningModule is in the first search result link for "pytorch lightning augmentation" -- https://lightning.ai/docs/pytorch/stable/notebooks/lightning_examples/augmentation_kornia.html

**Comment by calebrob6** [comment]
> I've been considering switching to only two parameters: preprocess (all splits) and augmentation (only train). This could be timed with a switch from Kornia to torchvision (if we decide to do this) to avoid a backwards-incompatible change.

I think this is a great idea!

> Further it isn't clear what exactly should and shouldn't be a kwarg.

The specific problem I was running into was when I wanted the datamodule to have the same kwarg as the dataset. E.g. I want to use a certain set of augmentations for band set "A" and a different set of augmentations for band set "B" -- if the datamodule has a parameter `bands` then I don't think it will by default go to the dataset.

**Comment by adamjstewart** [comment]
> The specific problem I was running into was when I wanted the datamodule to have the same kwarg as the dataset. E.g. I want to use a certain set of augmentations for band set "A" and a different set of augmentations for band set "B" -- if the datamodule has a parameter `bands` then I don't think it will by default go to the dataset.

In that case you would either skip the parameter in the datamodule (read from `kwargs['bands']`) or pass both the `bands` and `**kwargs` to the `dataset_class`. There are a few examples of this already, grep our builtin datamodules for `bands` for how this is handled.
